expect: Improve report when mock-spy matcher fails, part 2

pedrottimark opened the pull request:
## Summary

For mock-spy matchers whose criterion is number of calls or returns:

* Display matcher name in regular black instead of dim color
* Replace sentences with `Expected` and `Received` labels and number values

For negative matcher:

* with expected times, repeat `not` following `Expected` label
* without expected times, display `>= 1` following `Expected` label
* without expected times, display args or value as lines with one-based call number as label

**Questions** to reviewers

1. What do you think about change to consistent `PRINT_LIMIT = 3` instead of:

    ```js
    const CALL_PRINT_LIMIT = 3;
    const RETURN_PRINT_LIMIT = 5;
    ```

2. For **return** matchers, is the number of **calls** relevant information [EDIT: if it is different from the number of returns] to include on an additional line in report to trouble shoot situations when a test fails because of incorrect assumption about calls versus returns? (see picture in comment below)

**Residue**

* Factor out repeated code, especially to avoid `map` for spy calls unless negative matcher fails
* Edit test names to replace `in the error message` with `as received in matcher hint`
* Replace some `any` types, if possible
* Add tests to increase coverage

## Test plan

Updated 46 snapshots

| | long name | short name |
| ---: | :--- | :--- |
| 3 | `toHaveBeenCalled` | `toBeCalled` |
| 4 | `toHaveBeenCalledTimes` | `toBeCalledTimes` |
| 8 | `toHaveReturned` | `toReturn` |
| 8 | `toHaveReturnedTimes` | `toReturnTimes` |

See also pictures in following comments

Example pictures baseline at left and **improved at right**

pedrottimark commented:
To my eyes `>= 1` contrasts better than `> 0` with `0` and is more similar to explicit expected argument `1` in `toHaveBeenCalledTimes` or `toHaveReturned` matchers (see below)

<img width="1000" alt="toHaveBeenCalled false" src="https://user-images.githubusercontent.com/11862657/60742276-b976c900-9f3a-11e9-923e-042e216f24ab.png">

<img width="1000" alt="toHaveReturned false" src="https://user-images.githubusercontent.com/11862657/60742247-892f2a80-9f3a-11e9-9e92-da98d6c824a3.png">

Residue includes adding a test to cover the limit:

<img width="1000" alt="toHaveBeenCalled true" src="https://user-images.githubusercontent.com/11862657/60742318-e9be6780-9f3a-11e9-8ee4-4e739a722a2d.png">

<img width="1000" alt="toHaveReturned true" src="https://user-images.githubusercontent.com/11862657/60742254-93e9bf80-9f3a-11e9-9396-698bb5d9da94.png">

pedrottimark commented:
<img width="1000" alt="toHaveBeenCalledTimes false" src="https://user-images.githubusercontent.com/11862657/60742444-69e4cd00-9f3b-11e9-8c25-7e558475599a.png">

<img width="1000" alt="toHaveBeenCalledTimes true" src="https://user-images.githubusercontent.com/11862657/60742453-6ea98100-9f3b-11e9-9ff7-02f188c02da1.png">

I forgot to take baseline pictures of `toHaveReturnedTimes` which is similar to the above

pedrottimark commented:
The **call** number gives a **hint** that some invocations did not return (see Question 2)

<img width="1000" alt="toHaveReturned true 1 3" src="https://user-images.githubusercontent.com/11862657/60742791-c09ed680-9f3c-11e9-9075-0c7c9e70a0aa.png">

pedrottimark commented:
The following errors in CI also occurred when I ran `yarn jest` locally on Node 10.16.0:

1. ScriptTransformer transforms a file properly - ScriptTransformer transforms a file properly
2. generateEmptyCoverage generates an empty coverage object for a file without running it - generateEmptyCoverage generates an empty coverage object for a file without running it

I will wait for advice before I push update to `CHANGELOG.md` to save easy way to restart CI :)

SimenB commented:
Give it a rebase, it should fix ci 🙂

jeysal commented on packages/expect/src/spyMatchers.ts:
```diff
@@ -58,14 +52,24 @@ const createToBeCalledMatcher = (matcherName: string) =>
+          `Received number of calls: ${printReceived(count)}\n\n` +
+          calls
+            .reduce((lines: Array<string>, args: any, i: number) => {
+              if (lines.length < PRINT_LIMIT) {
```
Does this still look good for arguments that serialize to large strings?

pedrottimark replied:
Please see what you think about picture in https://github.com/facebook/jest/pull/8649#issuecomment-509270828

pedrottimark commented:
Here are pictures when received number of calls is not equal to returns

Updated `toHaveReturned` is at left and `toHaveReturnedTimes` is at right

Positive assertions:

<img width="1000" alt="toReturn positive" src="https://user-images.githubusercontent.com/11862657/60821883-69824700-a172-11e9-8938-23cff6d03e63.png">

Negative assertions:

<img width="1000" alt="toReturn negative" src="https://user-images.githubusercontent.com/11862657/60821876-65eec000-a172-11e9-99bd-6d9daec74679.png">

Updated 14 snapshots

| | long name | short name | |
| ---: | :--- | :--- | :--- |
| 4 | `toHaveReturned` | `toReturn` | |
| 3 | `toHaveReturnedTimes` | `toReturnTimes` | changed one test from negative to positive |

Here is picture of args that serialize to long strings as requested in https://github.com/facebook/jest/pull/8649#discussion_r300886037

<img width="1000" alt="toBeCalled positive serialize long" src="https://user-images.githubusercontent.com/11862657/60821840-58d1d100-a172-11e9-9c0f-24c5671d1dd3.png">

P.S. Danke, Tim: `erste` is adjective but `zuerst` is adverb, nicht wahr?

jeysal commented:
The arguments in the screenshot still look reasonably readable. I'm wondering though (for other PRs):
* How can we prevent broken output in extreme cases such as multiline strings (ellipsis abbreviations?)
* Does it make sense to use limited multiline printing if there are many arguments or an argument is an array/object with many properties

Also, should we perhaps print the arguments as `42, {}` instead of `[42, {}]` because users might think that the array means that it's one argument that is an array?

---

And yes you're right, *zuerst* is an adverb and *erster/erste/erstes* a numeral/adjective ;)

pedrottimark commented:
@jeysal Your review comments are sooo helpful

* For multiline strings, what do you think about ↵ `U+21b5` downwards arrow with corner leftwards.
* The `stringify` function has optional `maxDepth = 10` arg. What value do you think in this context?

The reason I printed args as array is possibility of call with no arguments. The `toHaveBeen*CalledWith` matchers will have `...expected` in matcher hint, but these matchers lack that hint, pardon the pun.

* If you think nothing following the label communicates clearly, then I am happy to write helper function to print values separated by commas, not enclosed in brackets.
* Or how about enclosed in parentheses instead of brackets? So `()` means no arguments.

pedrottimark commented:
Here are two variations on idea to enclose call args in **parentheses** instead of brackets.

Similar to example above one object arg enclosed in parentheses which have **black** color:

<img width="1000" alt="toHaveBeenCalled true paren 1" src="https://user-images.githubusercontent.com/11862657/60842306-cf39f780-a1a1-11e9-8a46-13f0d678d5ba.png">

Adapted to 3 args enclosed in parentheses and separated by commas which have **dim** color:

<img width="1000" alt="toHaveBeenCalled true paren 3" src="https://user-images.githubusercontent.com/11862657/60842292-c77a5300-a1a1-11e9-9276-56e63e007860.png">

When we get to `toHaveBeenLastCalledWith` and `toHaveBeenNthCalledWith` matchers, we will see whether or not it communicates clearly for arguments themselves to have dim color in report if matcher failed because received args match but **not** at the expected call.

jeysal commented:
> The reason I printed args as array is possibility of call with no arguments

For that case, maybe a placeholder `No arguments` would work and allow us to do it comma-separated :) Although using colors to signify that it's not an actual array is also an interesting idea

pedrottimark commented:
In report for negative `toHaveBeenCalled` matcher, call `printReceivedArgs` function:

* arguments have red color (as usual)
* enclose in **parentheses** and separate by comma-space (as usual) which have **dim** color

<img width="1000" alt="toHaveBeenCalled true 2 5 3" src="https://user-images.githubusercontent.com/11862657/60978986-eee73200-a2ff-11e9-9896-7fb15b1c3afa.png">

Although the example of unexpected call with **no args** requires me to look twice:

<img width="1000" alt="toHaveBeenCalled true 2 1 0" src="https://user-images.githubusercontent.com/11862657/60978974-eabb1480-a2ff-11e9-8c68-3ea8bf656260.png">

I think we will see in future improvement for `toHaveBeen*CalledWith` matchers that **contrast** between red arguments and dim punctuation will help compare received to expected

Updated 4 snapshots and changed 2 of them to illustrate 3 string args

Let’s defer multiline strings and object depth to formatting for data-driven diff

jeysal commented:
@SimenB @thymikee wdyt about the parentheses?
I think I'm still in favor of just commas and nothing at the start/end, with a fallback message for no args.

thymikee commented:
>I think I'm still in favor of just commas and nothing at the start/end, with a fallback message for no args.

I'm with @jeysal on that.

pedrottimark commented:
Thank you for constructive critique, my friends.

How do commas and `no args` look with normal black color?

<img width="1000" alt="toHaveBeenCalled true 3 5 3" src="https://user-images.githubusercontent.com/11862657/60988478-b18ba000-a311-11e9-9513-60bcc9655cac.png">

<img width="1000" alt="toHaveBeenCalled true 3 1 0" src="https://user-images.githubusercontent.com/11862657/60988468-acc6ec00-a311-11e9-991a-9a5dd72e6b8d.png">

thymikee commented:
Can we provide an information that 2 more calls were skipped?

```
1: "pierwszy"
2: "drugi"
3: "trzeci"
...skipped 2 more calls.
```

pedrottimark commented:
@thymikee Let’s wait to decide what to do until PR to improve report for `toHaveBeen*CalledWith` matchers, because they will probably have print-limited call lines with non-consecutive numbers.

pedrottimark commented:
@jeysal You’re welcome. Can you express your question in other words. I didn’t quite get it:

> Is there perhaps a similar message in spy matchers that we could adapt to "No args" to make it feel consistent?

jeysal commented:
> Can you express your question in other words. I didn’t quite get it:

"No args" looked somewhat odd to me and I wondered if we had a similar text somewhere that we could refer to for wording. But maybe it's "args" that looks a bit "vulgar" as an abbreviation. Doesn't matter much anyway 😄

pedrottimark commented:
@jeysal Aha, I see what you mean. It can be misunderstood as an error instead of neutral information.

Two existing strings will survive the “prose purge” in next PR for `toHave*ReturnedWith` matchers:

```sh
Expected value: 12                     [green color]
Received value:
             1: threw an error         [black color]
             2: has not returned yet   [black color]
             3: 13                     [red color]
```

Here is a first draft of a phrase that is parallel to their past tense grammar:

```sh
1: was called with no arguments        [black color]
```

jeysal commented:
+1 for called with no arguments :)
